kotlin/configure-subnetwork-trace/src/main/java/com/esri/arcgisruntime/sample/configuresubnetworktrace/MainActivity.kt

```diff
@@ -0,0 +1,360 @@
+package com.esri.arcgisruntime.sample.configuresubnetworktrace
+
+import android.os.Bundle
+import android.text.InputType
+import android.text.method.ScrollingMovementMethod
+import android.util.Log
+import android.view.View
+import android.widget.AdapterView
+import android.widget.ArrayAdapter
+import android.widget.Toast
+import androidx.appcompat.app.AlertDialog
+import androidx.appcompat.app.AppCompatActivity
+import com.esri.arcgisruntime.data.CodedValue
+import com.esri.arcgisruntime.data.CodedValueDomain
+import com.esri.arcgisruntime.loadable.LoadStatus
+import com.esri.arcgisruntime.utilitynetworks.UtilityAttributeComparisonOperator
+import com.esri.arcgisruntime.utilitynetworks.UtilityCategoryComparison
+import com.esri.arcgisruntime.utilitynetworks.UtilityElement
+import com.esri.arcgisruntime.utilitynetworks.UtilityElementTraceResult
+import com.esri.arcgisruntime.utilitynetworks.UtilityNetwork
+import com.esri.arcgisruntime.utilitynetworks.UtilityNetworkAttribute
+import com.esri.arcgisruntime.utilitynetworks.UtilityNetworkAttributeComparison
+import com.esri.arcgisruntime.utilitynetworks.UtilityTier
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceAndCondition
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceConditionalExpression
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceConfiguration
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceOrCondition
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceParameters
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceType
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraversability
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraversabilityScope
+import kotlinx.android.synthetic.main.activity_main.*
+
+
+class MainActivity : AppCompatActivity() {
+
+  private val TAG: String = MainActivity::class.java.simpleName
+
+  private val utilityNetwork by lazy {
+    UtilityNetwork(getString(R.string.naperville_electric_url))
+  }
+
+  private var initialExpression: UtilityTraceConditionalExpression? = null
+  private var sourceTier: UtilityTier? = null
+  private var sources: List<UtilityNetworkAttribute>? = null
+  private var operators: Array<UtilityAttributeComparisonOperator>? = null
+  private var startingLocation: UtilityElement? = null
+  private var values: List<CodedValue>? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    exampleTextView.movementMethod = ScrollingMovementMethod()
+
+    // create a utility network and wait for it to finish to load
+    utilityNetwork.loadAsync()
+    utilityNetwork.addDoneLoadingListener {
+      if (utilityNetwork.loadStatus == LoadStatus.LOADED) {
+        // create a list of utility network attributes whose system is not defined
+        sources = utilityNetwork.definition.networkAttributes.filter { !it.isSystemDefined }
+          .also { sources ->
```

Comment:
Personally I'm finding it difficult to follow the nesting here...is it necessary to use the also here? If not, it might be a bit more legible without it:

```// create a list of utility network attributes whose system is not defined
sources = utilityNetwork.definition.networkAttributes.filter { !it.isSystemDefined }

sources.sourceSpinner.apply {
  adapter = ArrayAdapter<String>(
    applicationContext,
    android.R.layout.simple_spinner_item,
    sources.map { it.name })

  onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
    override fun onItemSelected(
      parent: AdapterView<*>?,
      view: View?,
      position: Int,
      id: Long
     ) {
      (sources[sourceSpinner.selectedItemPosition])
     onComparisonSourceChanged(sources[sourceSpinner.selectedItemPosition])
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}
 }
```

Reply:
Hmm, I'm trying to work around it, but the .also scope provides a non-null `sources`

Reply:
That makes a lot of sense. I'm looking at it now and it's not so hard to follow, but maybe I'm just getting used to these scoping functions now. I haven't used .also too much yet

Reply:
I think after our discussion yesterday afternoon and looking at the activity as a whole, it makes sense. There's a pretty consistent pattern of
```
var myFish = Tuna().[scopingfxn] { fishy ->
  fishy.tank = this
  fishy.taste = Flavor.DELICIOUS
}
```
I guess scoping functions in kotlin let us block together all the kind of initialization code for objects during their creation? I can see the case for both the Java style of "each line does one action" versus this style of "group similar actions into one place". Maybe that's more of a mindset thing to get used to with kotlin?